src/ewah_benchmark.zig

```diff
@@ -0,0 +1,121 @@
+const std = @import("std");
+const assert = std.debug.assert;
+const ewah = @import("ewah.zig").ewah(usize);
+
+const BitSetConfig = struct {
+    words: usize,
+    run_length_e: usize,
+    literals_length_e: usize,
+};
+
+const samples = 100;
+const repeats: usize = 100_000;
+
+// Explanation of fields:
+// - "n": Number of randomly generate bitsets to test.
+// - "words": The length of the decoded bitset, in u64s.
+// - "run_length_e": The expected length of a run, ignoring truncation due to reaching the end of
+//   the bitset.
+// - "literals_length_e": Expected length of a sequence of literals.
+const configs = [_]BitSetConfig{
+    // primarily runs
+    .{.words = 640, .run_length_e = 10, .literals_length_e = 10},
+    .{.words = 640, .run_length_e = 100, .literals_length_e = 10},
+    .{.words = 640, .run_length_e = 200, .literals_length_e = 10},
+    // primarily literals
+    .{.words = 640, .run_length_e = 1, .literals_length_e = 100},
+};
+
+var prng = std.rand.DefaultPrng.init(42);
+
+pub fn main() !void {
+    const stdout = std.io.getStdOut().writer();
+
+    for (configs) |config| {
+        var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
+        defer arena.deinit();
+
+        var i: usize = 0;
+        var bitsets: [samples][]usize = undefined;
+        var bitsets_encoded: [samples][]align(@alignOf(usize)) u8 = undefined;
+        var bitsets_decoded: [samples][]usize = undefined;
+        var bitset_lengths: [samples]usize = undefined;
+        while (i < samples) : (i += 1) {
+            bitsets[i] = try make_bitset(&arena.allocator, config);
+            bitsets_encoded[i] = try arena.allocator.alignedAlloc(u8, @alignOf(usize), ewah.encode_size_max(bitsets[0]));
+            bitsets_decoded[i] = try arena.allocator.alloc(usize, config.words);
+        }
+
+        // Benchmark encoding.
+        const encode_timer = try std.time.Timer.start();
+        i = 0;
+        while (i < samples) : (i += 1) {
+            var j: usize = 0;
+            var size: usize = undefined;
+            while (j < repeats) : (j += 1) {
+                size = ewah.encode(bitsets[i], bitsets_encoded[i]);
+            }
+            bitset_lengths[i] = size;
+        }
+        const encode_time = encode_timer.read() / samples / repeats;
+
+        const decode_timer = try std.time.Timer.start();
+        // Benchmark decoding.
+        i = 0;
+        while (i < samples) : (i += 1) {
+            const bitset_encoded = bitsets_encoded[i][0..bitset_lengths[i]];
+            var j: usize = 0;
+            while (j < repeats) : (j += 1) {
+                _ = ewah.decode(bitset_encoded, bitsets_decoded[i]);
+            }
+        }
+        const decode_time = decode_timer.read() / samples / repeats;
+
+        i = 0;
+        while (i < samples) : (i += 1) {
+            assert(std.mem.eql(usize, bitsets[i], bitsets_decoded[i]));
```

Comment:
This is really valuable to have in the benchmark.

A simpler version of this (without bitset config) would be great to have in our tests as well, to make sure that we're roundtripping on fully random data also when unit testing. We could share some of this code between test/benchmark, but something simple as a unit test might not need any code sharing.

Just a very simple roundtrip as a unit test, a few loops, choosing different bitset length each loop, and fully random bits that we encode/decode and check for equality. Using small word sizes to make sure we test the max uniform/literal word count limits across the spectrum. That way, we don't need bitset config in the unit test (it's a good for benchmarking different distributions though).

Otherwise it might happen that somehow modifies something later, doesn't run the benchmark, the CI only runs the unit tests and the bug slips through.

Reply:
On second thought, perhaps the roundtripping we're doing in BlockFreeSet already covers this end-to-end.

We must just make sure that we're using small words so that we cover the limit cases.

Reply:
:+1: I'm going to add it. The encode->decode->encode case is well tested by both ewah & the free block set, but decode->encode->decode wasn't as well covered. (And that round trip is better-suited to fully -random fuzzing, since it can't accidentally encode invalid states).

Reply:
Awesome. Fuzzing for the win.